### ISSUE TITLE
Enable Flawless Installation On Windows Server Core 2016

### ISSUE THREAD

**Issue by softworkz** [MediaBrowser/Emby#3175]
This is a wrap-up of all the things learned so far about problems when emby is installed on a Windows Server Core 2016.

**Setup**

1. Installation of the Visual C++ 2015 Runtime fails without notice.
Potential remedies:

- Try to run the redist install in silent mode without GUI - if possible 
- Or use the VC merge modules, merge them into an msi setup and execute the msi as sub-setup (this will definetely work)

2. Setup should reject installing a 32bit version on Windows Server Core

This is to simplify the dll checking that will be required.
I wouldn't expect this to happen anyway, but [in this post](https://emby.media/community/index.php?/topic/12727-question-about-ffprobe-on-windows-server-core/) it is being talked about 32bit dlls...

** Check and fix missing DLLs**

The following dlls are required for ffmpeg:

- %WINDIR%\System32\msvfw32.dll
- %WINDIR%\System32\avicap32.dll

(crtdll and the mui resource files are not required)

And these dlls are required for SkiaSharp:

- %WINDIR%\System32\ddraw.dll
- %WINDIR%\System32\glu32.dll
- %WINDIR%\System32\opengl32.dll
- %WINDIR%\System32\dsound.dll
- %WINDIR%\System32\msvcp140.dll
- %WINDIR%\System32\vcruntime140.dll

The last two dlls are installed with the VC++ 2015 runtime. They can only be used to check for successful installation of the VC runtime, they cannot be installed just by copying (a lot of side-by-side binaries with appropriate policies need to be intalled)

For all other dlls it is sufficient to copy them into the System32 folder.

** Implementation **

Having the emby setup do that would be nice, but I don't think that this is going to happen ;-)

But I think we should do two things at least:

- Always check for presence of those dlls on startup and display a red warning message in the dashboard with some text or link to instructions about what needs to be done
- Also display this in the OOBE/Firsttime setup wizard

**Comment by softworkz** [comment]
## Where to get those dlls?

A number of users have copied the dlls from a Windows 10 installation.

But there's a better way to get the "right" dlls for the server version:

- Locate the installlation source from where you installed Windows Server
- If it's an ISO, mount it anywhere
- Look for "install.wim" - usually in the "sources" folder
- In fact this is just a zip file. Open it with a compression app of your choice
- The wim file contains several numbers (e.g. 1,2,3,4) representing different editions of the Os
- Check the "windows\system32" folder under each number until you find the files you're looking for and copy them to your Windows Core installation

**Comment by LukePulverenti** [comment]
I wonder if we could also just compile ffmpeg for windows with all dependencies self-contained.

**Comment by softworkz** [comment]
There is a msvfw32.lib but it is just a stub for dynamic linking to msvfw32.dll and suspect that it is similar for avicaps32. I guess, the reason is that these libraries are platform specific.

So no, we won't be able to compile this into ffmpeg, but in theory it might be possible to compile an ffmpeg version without these dependencies since we don't care about audio/video capabilities of the system we're running on. 
But I would need to investigate if this is high or low effort.

**Comment by softworkz** [comment]
One more note about the Skia installation:

I don't know why we stopped distributing the ImageMagick dlls. 
ImageMagick wouldn't require all those dlls like Skia, but: We would need to install the VC++ 2013 Runtime additionally (full package including MP stuff).

Again: All those runtimes could be merged into a single MSI package that could be installed silently.

**Comment by softworkz** [comment]
One more note: An MSI based setup for installing the runtimes would also solve the problem of #3027

**Comment by LukePulverenti** [comment]
Should we do anything here?

**Comment by softworkz** [comment]
Yes. All those and even a few more more issues could be addressed all at once with a single strike (MSI setup). Namely:
## For Win Server Core
- Installing the required VC runtimes (2015/2017) as part of the setup
- Enforcing specific bitness version on systems like core server
- For missing dlls on Win Server Core
    - Either provide a wizard that automates the process during setup by prompting to insert or mount the installation source or DVD during setup
    - Or much simpler: Only detect whether these dlls are present. If not, open a document giving instructions how to install those dlls
## General Windows Setup
- Windows Installer based setup (MSI)
- Installation of Emby Server as a native Windows service (without NSSM)
- Installing into the "Program Files " folder like normal applications (by default)
- Using a folder below "Program Data" for data like normal applications (by default)
- Choosing a custom installation path
- Choosing a custom data path


Did I forget anything?
